File: 56-merge-intervals/merge-intervals.cpp
class Solution {
public:
    vector<vector<int>> merge(vector<vector<int>>& intervals) {

        // sort them based on to their head so that the closer intervals remain closer.

        // [1,3] [2,6] [8,10] [15,18]

        // okay now

        sort(intervals.begin(),intervals.end());

        //okay now closer elements stay close

        // now traverse 
        vector<vector<int>>ans;

        for(auto interval: intervals)
        {
            if(ans.empty() || ans.back()[1]<interval[0])
            {
                ans.push_back(interval);
            }
            else
            {
                ans.back()[1]=max(ans.back()[1],interval[1]);
            }
        }

        return ans;


        
    }
};

File: 56-merge-intervals/merge-intervals.py
class Solution:
    def merge(self, intervals: List[List[int]]) -> List[List[int]]:


        # sort them as per their start


        intervals.sort(key=lambda x:x[0])

        ans=[]

        for start,end in intervals:

            if ans and ans[-1][1]>= start:
                ans[-1][1]=max(ans[-1][1], end)
            else:

                ans.append([start,end])

        return ans







        